src/lib/AnnotationFSM.ts

```diff
@@ -0,0 +1,63 @@
+import { AnnotationMode } from './AnnotationControls';
+
+export enum AnnotationState {
+    HIGHLIGHT = 'highlight',
+    HIGHLIGHT_TEMP = 'highlight_temp',
+    NONE = 'none',
+    REGION = 'region',
+    REGION_TEMP = 'region_temp',
+}
+
+export enum AnnotationInput {
+    CANCEL = 'cancel',
+    CLICK = 'click',
+    CREATE = 'create',
+    SUCCESS = 'success',
+    UPDATE = 'update',
+}
+
+export const modeStateMap = {
+    [AnnotationMode.HIGHLIGHT]: AnnotationState.HIGHLIGHT,
+    [AnnotationMode.NONE]: AnnotationState.NONE,
+    [AnnotationMode.REGION]: AnnotationState.REGION,
+};
+
+export const stateModeMap = {
+    [AnnotationState.HIGHLIGHT]: AnnotationMode.HIGHLIGHT,
+    [AnnotationState.HIGHLIGHT_TEMP]: AnnotationMode.HIGHLIGHT,
+    [AnnotationState.NONE]: AnnotationMode.NONE,
+    [AnnotationState.REGION]: AnnotationMode.REGION,
+    [AnnotationState.REGION_TEMP]: AnnotationMode.REGION,
+};
+
+export default class AnnotationFSM {
+    private currentState = AnnotationState.NONE;
+
+    public send = (input: AnnotationInput, mode: AnnotationMode): AnnotationMode => {
+        if (input === AnnotationInput.CLICK) {
```

Comment:
Will this transition from `highlight|region_temp` back to `none` if `CLICK` is the input? Also, from 'highlight|region' back to `none` if `CLICK` is the input?

Reply:
Yes, if `CLICK` is the input, and `none` is the mode.

Reply:
But if we make `AnnotationControls` dumber as you described in your above comment, mode will never be  `none` and the current FSM implementation won't work.